pkg/controllers/gateway/gateway_controller.go

```diff
@@ -351,6 +375,99 @@ func selectClusters(gateway gatewayv1beta1.Gateway) []string {
 	return []string{}
 }
 
+func (r *GatewayReconciler) DNSDeletion(ctx context.Context, gateway gatewayv1beta1.Gateway) error {
+	trafficAccessor := traffic.NewGateway(&gateway)
+	allHosts := trafficAccessor.GetHosts()
+
+	for _, host := range allHosts {
+		managedZone, subDomain, err := r.Host.GetManagedZoneForHost(ctx, host, trafficAccessor)
+		if err != nil {
+			return err
+		}
+		if managedZone == nil {
+			log.Log.Info("Managed zone not found cannot get DNS record")
+			return nil
+		}
+
+		log.Log.Info("getting DNS record associated with deleting gateway")
+		dnsRecord, err := r.Host.GetDNSRecord(ctx, subDomain, managedZone)
+		if err != nil && !k8serrors.IsNotFound(err) {
+			return err
+		}
+		if dnsRecord == nil {
+			log.Log.Info("dns record not found")
+			continue
+		}
+
+		log.Log.Info("deleting DNS record ", "DNSRecord:", dnsRecord.Name)
+		err = r.Delete(ctx, dnsRecord)
+		log.Log.Info("DNS record deleted ", "DNSRecord:", dnsRecord.Name)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *GatewayReconciler) cleanupCertificates(ctx context.Context, gateway *gatewayv1beta1.Gateway) error {
+	var hostsToRemove []string
+	var allGateways gatewayv1beta1.GatewayList
+
+	// get names of hosts for traffic object being deleted
+	for _, listener := range gateway.Spec.Listeners {
+		if string(*listener.Hostname) != "*." {
+			hostsToRemove = append(hostsToRemove, string(*listener.Hostname))
+		}
+	}
+	// list all the traffic objects
+	err := r.Client.List(ctx, &allGateways, client.InNamespace(gateway.Namespace))
```

Comment:
do we still need to do this?

Reply:
The idea is to check that none of the other Gateways is using this certificate before deleting it. The assumption I'm making here is that all the gateways that might reuse cert will reside in the same namespace (as it should represent a tenant)

Reply:
what we have so far are the following simplistic rules

```
How would work with wildcard hosts assigned to a gateway
A more specific host can be assigned to another Gateway e.g. i can use myapp.example.com in another Gateway even though *.example.com is specified in another gateway. This is expected behaviour when you are more specific with a hostname. Each more specific host would receive its own certificate. 
2 hostnames of the same value cannot be used in 2 different Gateways. If a hostname is required in more than 1 Gateway, use a single Gateway instead of 2. If there’s a use case for using a hostname in 2 different Gateways, we need to understand that better as we may be missing a feature
For the first Gateway implementation, each host will get its own certificate for that host. If a wildcard host is specified in a listener, a wildcard certificate (and dns record) will be used

```

Reply:
So with a rule that a host is unique to a gateway and each host gets its own certificate, I think we can now just remove the certificates associated with that gateway

Reply:
Pushed changes. It make my life much easier :laughing: